### ISSUE TITLE
Java SDK - Access Logical Entity Definitions

### ISSUE THREAD

**Issue by cashel1** [microsoft/CDM#256]
Working with the Contact entity within electronicMedicalRecords sample docs.  The contact.cdm.json, (consistent with the old model.json format) contains the logical definition of attributes unique to EMR and inherits attributes up thru the parent schema docs, crm/foundationCommon/applicationCommon/..etc. The CdmEnitiyDefinition.getAttributes() method returns the resolved object that incorporates all of the inherited class attributes.

I cannot locate any methods on CdmEntityDefinition or CdmEntityDeclarationDefinition that would return the logical definition unique to a manifest doc.  If available, would expect that the call to the CdmEntityDefinition.getExtendsEntity() would return a reference to the CRM module Contact entity.  Is this possible?

Also, can we expect that the new model, as provided, will enforce single inheritance? Our intent is to provide java functions at the highest point in the object hierarchy to support reuse.  Did note the doc reference that resolved objects should be used, which, if is the case, we would probably need to go with fat objects that included attributes that are not needed for higher level module usage. 

Thanks

**Comment by dingbx** [comment]
Hi cashel1,

For the first question, I think you mean that you want to get attributes from only the logical electronicMedicalRecords/contact.cdm.json entity. Did you call the getAttributes on the logical entity or the resolved entity?

For the single inheritance, an entity can only extends from one entity, but in your case looks like you want to extends from the same entity and deal only with the highest one.

**Comment by dingbx** [comment]
If you want to get only the attributes defined in your entity file, calling getAttributes on the logical entity should work. And fetchResolvedAttributes fetches all the attributes.

**Comment by cashel1** [comment]
Hello dingbx,

	First Quetion:
	
	I'm  using a CdmEntityDefinition (returned from a call to the manifest.getEntities()) and corpus 
 to return the CdmEntityDeclarationDefinition.
		
	for (final CdmEntityDeclarationDefinition entityDeclaration : manifest.getEntities()) {
		CdmEntityDefinition entSelected = cdmCorpus.<CdmEntityDefinition>fetchObjectAsync(entityDeclaration.getEntityPath(),manifest).join(); // Gets the entity object from the doc.
	}

	In making  the call to entityDeclaration.getAttributes(), It appears to be providing the resolved/inherited attributes
(first ones are createdOn, createdBy), so it looks like I'm using the 	resolved entity. I don't see any getLogical() type methods.
	
	Second Question:
	
	With the model.json approach, I was utilizing java single inheritance to create objects specific to a module.
 For example, CRM Instance inherits from foundationCommon/ApplicationCommon up to CDMEntity 
	as referenced in a classes schema definitions.  This inheritance approach allowed us to code java functions at the 'most general' level to support reuse in more specialized module instances that could then inherit these methods. With a resolved instance, the associated java object would loose this inheritatance/reuse functionality.  
	
	We are generating the java classes from the Cdm schema, so we could in some fashion, pull in or 
reference generalized set/get methods in the resolved class. 
	
	Suggestions are welcomed.

	Thanks

**Comment by Venkata1920** [comment]
Hi @cashel1, it looks like you're referencing a resolved manifest which would explain the entities also containing the resolved attributes. Is that the case? If not, could you share your workflow / code snippet from manifest creation to fetching the entity definitions?

**Comment by cashel1** [comment]
Hello Venkata1920
I am using the manifest included in the sample 1-read-manifest sample (CDM-master->samples->1-read-manifest->sample-data) which is default.manifest.cdm.json
I am accessing the corpus and manifest using the provided sample java program, so there is no customization at this point.

Comparing this manifest to the one in the schemaDocuments->core.....->electionicMedicalRecords, it does appear that the sample manifest has gone thru some form of compilation.  If so, can a subset of the electronicMedicalRecords be used against the samples to generate logical entities?

**Comment by Venkata1920** [comment]
Hi @cashel1, if you're using the sample code as is, then that is the issue. The sample code overrides the manifest path for CDM standard manifests since we want all the sample data to be locally available. Please comment out or remove that code path (look for comment about fake "cdm") and then retry. You should be seeing the logical definitions of the entities as expected. Please get back if you still see issues after that.

**Comment by cashel1** [comment]
Hello Venkata1920

Thanks for the feedback on  accessing the logical schema attributes.  Below is the relevant portion of the read-manifest sample code.
Utilizing this code, with the removal of the localStandardsAdapter per your suggestion, I am still receiving all of the resolved attributes.  

I modified the code (note the CHANGE notations), to point to the CDM-Master/.../foundationCommon and its manifest. The manifest.getEntities() returns the foundationCommon entities, but  I am not able to access the attribute  values vis this approach.  The attributes are returned as a CdmAttributeGroupReference, and I am not able to locate any methods to return the the attribute/trait values.

Thanks for your assistance.  


final CdmCorpusDefinition cdmCorpus = new CdmCorpusDefinition();

// set callback to receive error and warning logs.
cdmCorpus.setEventCallback((level, message) -> {
	System.out.println(message);
}, CdmStatusLevel.Warning);

// ---------------------------------------------------------------------------------------------
// Configure storage adapters and mount them to the corpus.

// We want our storage adapters to point at the local manifest location and at the example public standards.
final String pathFromExeToExampleRoot = "../../";
// CHANGE point to CDM-Master as root
//final String pathFromExeToExampleRoot = "../../../";

LocalAdapter localDataAdapter  = new LocalAdapter(pathFromExeToExampleRoot + "1-read-manifest/sample-data");
// CHANGE point to foundationCommon
//LocalAdapter localDataAdapter  = new LocalAdapter(pathFromExeToExampleRoot + "schemaDocuments/core/applicationCommon/foundationCommon");

cdmCorpus.getStorage().mount("local",localDataAdapter);


// 'local' is our default namespace.
// Any paths that start navigating without a device tag (ex. 'cdm')
// will just default to the 'local' namespace.
cdmCorpus.getStorage().setDefaultNamespace("local");


// Storage adapter pointing to the example public standards.
// This is a fake 'cdm'; normally the CDM Standards adapter would be
// used to point at the real public standards.
// Mount it as the 'cdm' device, not the default,
// so that we must use "cdm:<folder-path>" to get there.

// CHANGE COMMENTED OUT per suggestion
//LocalAdapter localStandardsAdapter  = new LocalAdapter(pathFromExeToExampleRoot + "example-public-standards");
//cdmCorpus.getStorage().mount("cdm",localStandardsAdapter);

String manifestPath = "default.manifest.cdm.json";
// CHANGE point to the foundationCommon manifest
//String manifestPath = "foundationCommon.manifest.cdm.json";

CdmManifestDefinition manifest = cdmCorpus.<CdmManifestDefinition>fetchObjectAsync(manifestPath).join();
if (manifest == null) {
	System.err.println("Unable to load manifest " + manifestPath + ". Please inspect error log for additional details.");
	return;
}


for (final CdmEntityDeclarationDefinition entityDeclaration : manifest.getEntities()) {
	
	System.out.println("Reading the entity schema for " + entityDeclaration.getEntityName() + " located at " + entityDeclaration.getEntityPath());

	// From the path to the entity, get the actual schema description.
	// Take the local relative path in this doc and make sure it works.
	CdmEntityDefinition entity = cdmCorpus.<CdmEntityDefinition>fetchObjectAsync(entityDeclaration.getEntityPath(),manifest).join(); // Gets the entity object from the doc.
	
	for (final CdmAttributeItem attribute : entity.getAttributes()) {
		if (attribute instanceof CdmTypeAttributeDefinition) {
			final CdmTypeAttributeDefinition typeAttributeDefinition = (CdmTypeAttributeDefinition) attribute;
			System.out.println(typeAttributeDefinition.getName());
		} else if (attribute instanceof CdmAttributeGroupReference)  {	
			CdmAttributeGroupReference groupRef = (CdmAttributeGroupReference) attribute;
			System.out.println("review");
			// ISSUE, for foundationCommon Account, first attribute should be the defaultPriceLevel - no methods available for accessing atb type info
		} else {
			System.out.println("what is it");
		}
	}
	
}

**Comment by Venkata1920** [comment]
hi @cashel1, when initializing the LocalAdapter, you need to provide the folder that acts as the root to resolve all references. Given that the local adapter is pointing to foundationCommon, it will not be able to reference anything that needs to come from foundations.cdm.json. So, you will need to provide the path only schemaDocuments to the LocalAdapter ctor. For the manifest path, provide the relative path from there [string manifestPath = local:/core/applicationCommon/foundationCommon/crmCommon/accelerators/healthCare/electronicMedicalRecords/electronicMedicalRecords.manifest.cdm.json]

**Comment by cashel1** [comment]
Hello Venkata1920,

Thanks for the suggestions.  I was successful in returning all of the entities in the emr module. However, it appears that
I need to resolve the entities as the getAttributes() returns CdmAttributeGroupReference object vs a CdmTypeAttributeDefinition as provided in the sample code.   

Assuming this is the case, I attempted to use the following without success.
 CdmManifestDefinition manifestResolved =  manifest.createResolvedManifestAsync("default", "").get();

At this point I'm just interested in generating the resolved entities.

Thanks 

final CdmCorpusDefinition cdmCorpus = new CdmCorpusDefinition();

		// set callback to receive error and warning logs.
		cdmCorpus.setEventCallback((level, message) -> {
			System.out.println(message);
		}, CdmStatusLevel.Warning);

		final String pathFromExeToExampleRoot = "../../../";
		
		LocalAdapter localDataAdapter  = new LocalAdapter(pathFromExeToExampleRoot + "schemaDocuments");
		
		cdmCorpus.getStorage().mount("local",localDataAdapter);
		
		cdmCorpus.getStorage().setDefaultNamespace("local");

		String manifestPath = "local:/core/applicationCommon/foundationCommon/crmCommon/accelerators/healthCare/electronicMedicalRecords/electronicMedicalRecords.manifest.cdm.json";
		
		CdmManifestDefinition manifest = cdmCorpus.<CdmManifestDefinition>fetchObjectAsync(manifestPath).join();

		// at this point, have all the EMR entities, unable to revolve them to generate CdmTypeAttributeDefinition's 
			
		try {	
	
			for (final CdmEntityDeclarationDefinition entityDeclaration : manifest.getEntities()) {

				// From the path to the entity, get the actual schema description.
				// Take the local relative path in this doc and make sure it works.
				final CdmEntityDefinition entity = cdmCorpus.<CdmEntityDefinition>fetchObjectAsync(entityDeclaration.getEntityPath(), manifest).join(); // Gets the entity object from the doc.

				for (final CdmAttributeItem attribute : entity.getAttributes()) {
					if (attribute instanceof CdmTypeAttributeDefinition) {
						final CdmTypeAttributeDefinition typeAttributeDefinition = (CdmTypeAttributeDefinition) attribute;
						System.out.println(typeAttributeDefinition.getName());
					} else if (attribute instanceof CdmAttributeGroupReference)  {	
						CdmAttributeGroupReference groupRef = (CdmAttributeGroupReference) attribute;
						System.out.println("review");
						// ISSUE, for foundationCommon Account, first attribute should be the defaultPriceLevel - no methods available for accessing atb type info
					} else {
						System.out.println("what is it");
					}
				}

			}
		} catch (Exception e) {
			System.out.println(e.toString());
		}

**Comment by cashel1** [comment]
Hello Venkata1920,
Was able to get the entity resolution working by utilizing C# example from 2-create-manifest/sample-data.  Required restoring the fake 'cdm' I had removed per your suggestion:
  // CHANGE COMMENTED OUT per suggestion
  LocalAdapter localStandardsAdapter  = new LocalAdapter(pathFromExeToExampleRoot + "example-public-standards");
  cdmCorpus.getStorage().mount("cdm",localStandardsAdapter);

and adding the manifest build code from the create-manifest example.

I did receive the following deprecation warning.

CdmEntityDefinition | Resolution guidance is being deprecated in favor of Projections. https://docs.microsoft.com/en-us/common-data-model/sdk/convert-logical-entities-resolved-entities#projection-overview. | createResolvedEntityAsync | cdm:/core/applicationCommon/foundationCommon/crmCommon/accelerators/healthCare/electronicMedicalRecords/Account.cdm.json/Account

Thanks for all your help in resolving this.
